misc/kafka-util/summarize_topics.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+"""summarize_topics
+
+Basic script to print messages count and number of bytes within a Kafka topic
+"""
+
+import argparse
+import fnmatch
+import time
+
+import kafka
+
+
+def summarize_topic(args, topic):
+    """Read messages from a topic and print basic information about the topic."""
+
+    consumer = kafka.KafkaConsumer(
+        topic,
+        auto_offset_reset="earliest",
+        consumer_timeout_ms=1000,
+        bootstrap_servers=[f"{args.kafkahost}:{args.port}"],
+        enable_auto_commit=True,
+    )
+
+    start = time.time()
+
+    num_messages = 0
+    key_bytes = 0
+    value_bytes = 0
+    for message in consumer:
+        num_messages += 1
+        key_bytes += len(message.key) if message.key else 0
+        value_bytes += len(message.value) if message.value else 0
+
+    seconds_elapsed = time.time() - start
+    print(f"{topic},{num_messages},{key_bytes},{value_bytes},{seconds_elapsed:.1f}s")
+
+
+def summarize_topics(args):
+    """Read messages from topics matching prefix and print basic information about the topic."""
+
+    consumer = kafka.KafkaConsumer(bootstrap_servers=[f"{args.kafkahost}:{args.port}"])
+    topics = sorted(fnmatch.filter(consumer.topics(), args.topic_filter))
+
+    print("Topic,NumMessages,KeyBytes,ValueBytes,PythonConsumerTimeElapsed")
+    for topic in topics:
+        summarize_topic(args, topic)
+
+
+def main():
+    """Parse arguments and print topic summaries."""
```

Comment:
sort of a nit, but this script is structured from helper to entry point, such that the most in-the-weeds function is the one you read first upon seeing this file, and you have to go to the end of the file to get a sense of the structure.

IMO this would be a bit easier to understand if the functions were ordered:

* main
* summarize_topics
* summarize_topic

that would also match the general structure of the executables in this repo.

Reply:
If that's the structure that we prefer, I'm happy to change my habits.

I've always written things "bottom up", so that when you're reading `summarize_topics`, `summarize_topic` has already been defined. This is also a holdover from writing a lot of C++ code that rarely declared methods and instead only defined them (and therefore had to be written bottom up).

What do you think? I also wrote the code in #4885 bottom-up.